ambry-tools/src/main/java/com.github.ambry/clustermap/HelixBootstrapUpgradeTool.java

```diff
@@ -0,0 +1,667 @@
+      instances[i] = instanceName;
+      InstanceConfig instanceConfig = dcAdmin.getInstanceConfig(clusterName, instanceName);
+      Map<String, String> diskInfo = instanceConfig.getRecord().getMapField(replica.getMountPath());
+      String replicasStr = diskInfo.get(REPLICAS_STR);
+      replicasStr += replica.getPartition().getId() + REPLICAS_DELIM_STR;
+      diskInfo.put(REPLICAS_STR, replicasStr);
+      instanceConfig.getRecord().setMapField(replica.getMountPath(), diskInfo);
+      if (sealed) {
+        List<String> currentSealedPartitions = instanceConfig.getRecord().getListField(SEALED_STR);
+        List<String> newSealedPartitionsList = new ArrayList<>(currentSealedPartitions);
+        newSealedPartitionsList.add(partitionName);
+        instanceConfig.getRecord().setListField(SEALED_STR, newSealedPartitionsList);
+      }
+      // @todo: uncomment when we move to Helix 0.6.7.
+      // dcAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+    }
+    return instances;
+  }
+
+  /**
+   * Helper method to get the list of {@link ReplicaId} of all replicas of a partition in the given datacenter.
+   * @param partition the partition of interest.
+   * @param dcName the datacenter to which the returned replicas should belong.
+   * @return a list of {@link ReplicaId} of all replicas of the given partition in the given datacenter.
+   */
+  private List<ReplicaId> getReplicasInDc(Partition partition, String dcName) {
+    // returns a copy unlike getReplicas()
+    List<ReplicaId> replicaList = partition.getReplicaIds();
+    ListIterator<ReplicaId> iter = replicaList.listIterator();
+    while (iter.hasNext()) {
+      if (!iter.next().getDataNodeId().getDatacenterName().equals(dcName)) {
+        iter.remove();
+      }
+    }
+    return replicaList;
+  }
+
+  /**
+   * Get the instance name string associated with this data node in Helix.
+   * @param dataNode the {@link DataNodeId} of the data node.
+   * @return the instance name string.
+   */
+  private static String getInstanceName(DataNodeId dataNode) {
+    return dataNode.getHostname() + "_" + dataNode.getPort();
+  }
+
+  /**
+   * Validate that the information in Helix is consistent with the information in the static clustermap; and close
+   * all the admin connections to ZK hosts.
+   */
+  private void validateAndClose() {
+    try {
+      verifyEquivalencyWithStatic(staticClusterMap.hardwareLayout, staticClusterMap.partitionLayout);
+    } finally {
+      for (HelixAdmin admin : adminForDc.values()) {
+        admin.close();
+      }
+    }
+  }
+
+  /**
+   * Verify that the information in Helix and the information in the static clustermap are equivalent.
+   * @param hardwareLayout the {@link HardwareLayout} of the static clustermap.
+   * @param partitionLayout the {@link PartitionLayout} of the static clustermap.
+   */
+  private void verifyEquivalencyWithStatic(HardwareLayout hardwareLayout, PartitionLayout partitionLayout) {
```

Comment:
may be rename to "verifyEquivalencyWithStaticClusterMap" or "verifyEquivalencyWithStaticCS"